### PR TITLE
Node Graph Improvements

### PR DESCRIPTION
**Pull request by PlasmaDev5**
Backport of a few style features for nodes graphs.

Background grid and Subway nodes
image

Node graph debug animation. Not currently used in any meaningful way but there ready for when debugging becomes a thing
NodeDebug

### CODE REVIEW

File: Code/Tools/Libs/GuiFoundation/NodeEditor/Implementation/Connection.cpp
Comment (C-Core) [inline Code/Tools/Libs/GuiFoundation/NodeEditor/Implementation/Connection.cpp:190]: There also needs to be something on the connection it self to tell whether it should show the debugging effect. Can be added later though.

File: Code/Tools/Libs/GuiFoundation/NodeEditor/Implementation/Connection.cpp
Comment (C-Core) [inline Code/Tools/Libs/GuiFoundation/NodeEditor/Implementation/Connection.cpp:204]: This looks quite involved. Can you explain what a "subway path" does?

File: Code/Tools/Libs/GuiFoundation/NodeEditor/NodeScene.moc.h
Comment (C-Core) [inline Code/Tools/Libs/GuiFoundation/NodeEditor/NodeScene.moc.h]: I don't think we should change the default. There should be a view option or something like that where the user can toggle between bezier and subway lines.

File: Code/Tools/Libs/GuiFoundation/NodeEditor/Implementation/Connection.cpp
Comment (PlasmaDev5) [reply Code/Tools/Libs/GuiFoundation/NodeEditor/Implementation/Connection.cpp:190]: Yea thats the intention at the moment. I haven't added any debug logic myself yet but it made sense to include the visuals in the changes i was making

File: Code/Tools/Libs/GuiFoundation/NodeEditor/Implementation/Connection.cpp
Comment (PlasmaDev5) [reply Code/Tools/Libs/GuiFoundation/NodeEditor/Implementation/Connection.cpp:204]: Ill add a comment for it.
It draws the same way you would see on many subway maps. this means only straight and diagonal lines to reach the end goal. It also handles connections to nodes to the left of the output 

Subway

File: Code/Tools/Libs/GuiFoundation/NodeEditor/NodeScene.moc.h
Comment (PlasmaDev5) [reply Code/Tools/Libs/GuiFoundation/NodeEditor/NodeScene.moc.h]: I feel like ideally it should be something in the editor settings along with path width and other things like that. But i havent looked into doing that kind of thing yet. The change there was more for testing and can be reverted but i feel it gives a much more polished look for now until there is a setting for it.

Comment (C-Core) [reply Code/Tools/Libs/GuiFoundation/NodeEditor/NodeScene.moc.h]: I agree that it looks nice. The reason I'm hesitant to change the default is that you will position your nodes depending on how the connections behave to untangle the mess a bit. If we simply change the default existing graphs might look weird because they have been layouted with a different connection behavior. I will play around with these changes over the weekend and test them on a few more complex graphs.

**Comment by jankrassnigg** [comment]
Very cool! Thanks a lot! I'll leave the review up to @C-Core though.

File: Code/Tools/Libs/GuiFoundation/NodeEditor/Implementation/Connection.cpp
Comment (C-Core) [inline Code/Tools/Libs/GuiFoundation/NodeEditor/Implementation/Connection.cpp]: This additional moveTo is unnecessary and causes artifacts since the lines are not properly connected at this point.

File: Code/Tools/Libs/GuiFoundation/NodeEditor/Implementation/NodeView.cpp
Comment (C-Core) [inline Code/Tools/Libs/GuiFoundation/NodeEditor/Implementation/NodeView.cpp:139]: While it looks nice this unfortunately costs way too much performance and makes the whole node view react very poorly while panning especially when zoomed out a bit. You can leave the code in there but put an if(false) or something like that around it so we can add an editor setting for it in the future.

File: Code/Tools/Libs/GuiFoundation/NodeEditor/NodeScene.moc.h
Comment (C-Core) [reply Code/Tools/Libs/GuiFoundation/NodeEditor/NodeScene.moc.h]: I gave it a shot with a couple of complex graphs. It is actually not too bad, but I would still not change the default as it can be confusing in some existing graphs. I can look into adding the corresponding editor setting after this PR has been merged.

File: Code/Tools/Libs/GuiFoundation/NodeEditor/Implementation/NodeView.cpp
Comment (PlasmaDev5) [reply Code/Tools/Libs/GuiFoundation/NodeEditor/Implementation/NodeView.cpp:139]: Potentially another option in the graph settings as background grids are fairly standard these days for node tools

File: Code/Tools/Libs/GuiFoundation/NodeEditor/Implementation/Connection.cpp
Comment (PlasmaDev5) [reply Code/Tools/Libs/GuiFoundation/NodeEditor/Implementation/Connection.cpp]: Fixed

**Review by C-Core — APPROVED** [review]

File: Code/Tools/Libs/GuiFoundation/NodeEditor/Implementation/NodeView.cpp
Comment (PlasmaDev5) [reply Code/Tools/Libs/GuiFoundation/NodeEditor/Implementation/NodeView.cpp:139]: Do we have knowledge of the zoom level. we could adapt the graph drawing based on zoom lower draw cost when navigating large graphs

Comment (PlasmaDev5) [reply Code/Tools/Libs/GuiFoundation/NodeEditor/Implementation/NodeView.cpp:139]: @C-Core been playing around with this a little. Not the cleanest solution but improves the performance a tone. In future i ideally id want something that has multiple resolutions scales, so at each 0.25 step the graph looks the same but scaled for that resolution if that makes sense

```
  if(m_ViewScale.manhattanLength() > 1.0)
  {
    QPen pfine(plToQtColor(plColorScheme::GetColor(plColorScheme::Black, 9)), 1.0);

    painter->setPen(pfine);
    DrawGrid(painter, 15);
  }

  if(m_ViewScale.manhattanLength() > 0.1)
  {
    double scale = m_ViewScale.manhattanLength() < 0.25 ? 150.0 : 300.0;

    QPen p(plToQtColor(plColorScheme::GetColor(plColorScheme::Gray, 1)), 1.0);

    painter->setPen(p);
    DrawGrid(painter, scale);
  }
  ```

**Pull request merged**
